### ISSUE TITLE
likelihood fucntion problem

### ISSUE THREAD

**Issue by mavino** [veg/hyphy#858]
I am trying to run a batch file to get a likelihood function for many different files with script 'fit_codon_model'.
The analysis starts and performs well giving me the likelihood function for the first file and then it stops.
This is the error
```
{
 "00":"Universal",
 "01":"/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00002.fasta",
 "02":"Global w/variation",
 "03":"Gamma+Inv",
 "04":"4",
 "05":"012345",
 "06":"/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00002.nwk",
 "07":"/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00002.LK"
}Error:

Master node received an error:Attempted to delete a data set filter which is still being referenced by a likelihood function

Function call stack
1 : Read Data Set codon_ds from file PROMPT_FOR_FILE
	Standard input redirect:
		02 : Global w/variation
		03 : Gamma+Inv
		04 : 4
		05 : 012345
		06 : /home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00002.nwk
		07 : /home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00002.LK
		-------
2 : ExecuteAFile from file script using basepath /home/marianoavino/Desktop/staph/. reading input from redirect
{
 "00":"Universal",
 "01":"/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00002.fasta",
 "02":"Global w/variation",
 "03":"Gamma+Inv",
 "04":"4",
 "05":"012345",
 "06":"/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00002.nwk",
 "07":"/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00002.LK"
}
-------
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD 
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```

This is my batch file:
```
// path to script you want to run
// "/usr/local/lib/hyphy" +
script = "/home/marianoavino/Desktop/staph/fit_codon_model.bf";


// path to a text file containing absolute paths to each data file
fscanf("/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/fasta.csv", "Lines", paths);


for (_path = 0; _path < Columns(paths); _path = _path+1) {
     // compose output file name
     outpath = paths[_path]^{{"\.fasta"}{".LK"}};

     // list of commands from interactive HyPhy session
     redirect = {};
     redirect["00"] = "Universal";
     redirect["01"] = paths[_path];  // path to input file
     redirect["02"] = "Global w/variation";
     redirect["03"] = "Gamma+Inv";
     redirect["04"] = "4";
     redirect["05"] = "012345";
     redirect["06"] = paths[_path]^{{"\.fasta"}{".nwk"}};  // input tree
     redirect["07"] = outpath;  // output LF

     fprintf(stdout, redirect);
     ExecuteAFile(script, redirect);
}
```

I get the same kind of error either I run `HYPHYMPI batch` and `HYPHYMP batch` on command line.

My HYPHY version is HYPHY 2.3.14.20181026beta(MP) or HYPHY 2.3.14.20181026beta(MPI).

Any idea?
Thank you very much.

**Comment by spond** [comment]
Dear @mavino,

In your `fit_codon_model.bf` add a statement **at the end of the file** that looks like this

```
DeleteObject (lf_id); 
// replace lf_id with the actual name of the LikelihoodFunction you are creating/optimizing
```

This will clear out the relevant objects and prevent the error you see from occurring.

Best,
Sergei

**Comment by mavino** [comment]
Dear @spond...
Yes, it worked adding the line you suggested.
Thank you very much.
Best regards,
Mariano

**Comment by mavino** [comment]
I am sorry if I open this issue again but something came up again at it might be related to the old one.
So, I have run my job in batch according modifications suggested by @spond (see above) and it worked fine with 256 out of 259 samples.
The three not working samples produced blank files.
I separately rerun (with same code as above) one of them and I got this error:

```
Error:
Internal error, dumping the offending likelihood function to /tmp/hyphy.dumpInternal error in _LikelihoodFunction::ConjugateGradientDescent. The function evaluated at current parameter values [-1.01348e+06] does not match the last recorded LF maximum [-14221.9]

Function call stack
1 : Optimize storing into, res, the following likelihood function:codon_lf ; 
	Standard input redirect:
		Empty Associative List-------
2 : ExecuteAFile from file script using basepath /home/marianoavino/Desktop/staph/Test/. reading input from redirect
{
 "00":"Universal",
 "01":"/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00524.fasta",
 "02":"Global w/variation",
 "03":"Gamma+Inv",
 "04":"4",
 "05":"012345",
 "06":"/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00524.nwk",
 "07":"/home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00524.LK"
}
-------

Check errors.log for details on execution errors.

Check messages.log details of this run.
```

Any idea?
Thank you very much.

**Comment by spond** [comment]
Dear @mavino,

I would like to resolve this issue, but I do need you to send me the files (SAOUHSC_00524.fasta and SAOUHSC_00524.nwk) that trigger the issue so that I can replicate it.

Many thanks,
Sergei

**Comment by mavino** [comment]
Dear @spond;
here you are:
[SAOUHSC_00524.zip](https://github.com/veg/hyphy/files/2663443/SAOUHSC_00524.zip)
Thank you very much.
Best,
Mariano

**Comment by spond** [comment]
Dear @mavino,

I was not able to replicate the error with the 2.4 (alpha) development build; but I did notice one other thing.

The alignment you sent me consisted of about 95% duplicate sequences and was quite large (~8000 sequences). For selection analyses, the inclusion of duplicate sequences contributes essentially no information to the analysis and can **greatly** increase run time. We recommend that you `de-duplicate` sequences in your input alignments, i.e. only keep one _copy_ of each duplicate sequence. 

I am attaching a simple batch file to do that, with the added benefit of trimming all the corresponding leaves from a large phylogeny. I suggest you do this step prior to fitting any selection-detection models in HyPhy.

When running your large file on the  trimmed version of your dataset (reduced to under 500 sequences and cutting runtime by an order of magnitude), I obtained the following result (abbreviated, of course)

```
...
### Rate class analyses summary
*  932 branches with **1** rate classes
*  3 branches with **2** rate classes

### Improving parameter estimates of the adaptive rate class model
* Log(L) = -11253.48, AIC-c = 26299.89 (1890 estimated parameters)

...

### Adaptive branch site random effects likelihood test 
Likelihood ratio test for episodic diversifying positive selection at Holm-Bonferroni corrected _p =   0.0500_ found **0** branches under selection among **935** tested.

```

It is worth noting that aBSREL loses power for larger trees with short branches (because of multiple test correction), so this result is not too surprising. In fact, I'd not at all be surprised if you **very rarely** found evidence of selection on a single branch in an 8,000 taxon tree, simply because there is only so much signal you get from a single branch, and the stringency of multiple testing correction gets tighter with larger trees.

Best,
Sergei

[DuplicateTrimmer.txt](https://github.com/veg/hyphy/files/2682690/DuplicateTrimmer.txt)

**Comment by mavino** [comment]
Dear @spond thank you very much, 
that was very kind of you, especially to make me DuplicateTrimmer.txt...
on that note I am getting the following error though:
```
HYPHYMPI DuplicateTrimmer.txt 

Select a sequence alignment file (`/home/marianoavino/Desktop/staph/Test/`) /home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00524.fasta
> Read an alignment on 7878 sequences and 3549 sites.

Please select a tree file for the data: (`/home/marianoavino/Desktop/staph/Test/`) /home/marianoavino/Desktop/staph/inframe/edited/prunedfasta/SAOUHSC_00524.nwk

Found **469** unique sequences
Removing **7409** duplicate sequences and corresponding leaves in the tree
Error:

Master node received an error:An invalid argument (not a string) supplied to _TreeTopology::RemoveANode

Function call stack
1 : T-utility.Values(to_delete)
-------
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD 
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
```
Best
Mariano

**Comment by spond** [comment]
Dear @mavino,

Sorry about this -- I wrote the script to work with v2.4alpha. You can get it from the `develop` branch if you''d like.

Best,
Sergei

**Comment by mavino** [comment]
Dear @spond,
no problem about it.
I had already a script that handled removing duplicates from alignments and another one that pruned the corresponding newick tree file.

So I removed the duplicated sequences and I reran the analysis and I was able to get my original analysis done.
It must have been that my HYPHY version could not handled so many duplicated sequences.
Thank you very much for the hint and all the help.
Best,
Mariano
